### PR TITLE
amp-script: Support local scripts via "script" attribute

### LINKED ISSUE #22612: [amp-script] Support inline JS
Not sure if this came up already (at least I couldn't find anything): it looks like amp-script currently only supports embedding external JS via the `src` attribute. It'd be great to be able to define custom scripts inline:

```html
 <amp-script layout=container src="/examples/amp-script/hello-world.js">
    <button id="greeting">Say hello</button>
    <script>
    const button = document.getElementById('greeting')
    ...
    </script>
  </amp-script>
``` 

I see a few advantages for this:

1. This would greatly simplify documenting amp-script as samples would no longer need two files. 
2. Faster loading time (we probably should restrict script size for these cases even further).

### PR DESCRIPTION
**Pull request by choumx**
Fixes #22612. 

Open to suggestions for an attribute name other than "script" which is a tad overloaded. Was following convention set by "template" attribute.

```html
<amp-script script=myScript width=300 height=100>
</amp-script>

<script type=text/amp-script id=myScript>
</script>
```

Note these are only reference by id rather than inline (children) since I wanted to avoid the hydration filtering issue.

### CODE REVIEW

File: extensions/amp-script/validator-amp-script.protoascii [hunk 1]
```diff
@@ -24,17 +24,42 @@ tags: {
...
-
+tags: {  # <amp-script> (local script)
+  html_format: AMP
+  tag_name: "SCRIPT"
+  spec_name: "amp-script extension local script"
+  requires_extension: "amp-script"
+  attrs: {
+    name: "type"
+    mandatory: true
+    value: "text/amp-script"
```
[the remark this answers is not in the document]
Comment (choumx) [reply extensions/amp-script/validator-amp-script.protoascii]: Any `script[type]` value that's not a JS MIME type is treated as a data block. https://html.spec.whatwg.org/multipage/scripting.html#attr-script-type

text/amp-script` is not a valid MIME type but very unlikely to collide with a future MIME type.

[the remark this answers is not in the document]
Comment (choumx) [reply extensions/amp-script/validator-amp-script.protoascii]: Or we could use `type=text/plain` and use a non-standard attribute to mark its usage e.g.

```html
<script type=text/plain target=amp-script>
```

Happy with either and open to suggestions on the attribute name.

File: extensions/amp-script/0.1/test/validator-amp-script-not-origin-trial.out [hunk 2]
```diff
@@ -34,4 +34,4 @@ FAIL
 |    <amp-script layout=container src="https://example.com/foo.js"><div class="root"><p>Some text to hydrate.</p></div></amp-script>
 |  </body>
 >>       ^~~~~~~~~
-amp-script/0.1/test/validator-amp-script-not-origin-trial.html:34:6 The tag 'amp-experiment-token' is missing or incorrect, but required by 'amp-script'. (see https://amp.dev/documentation/components/amp-script) [AMP_TAG_PROBLEM]
+amp-script/0.1/test/validator-amp-script-not-origin-trial.html:34:6 The tag 'amp-experiment-token' is missing or incorrect, but required by 'amp-script'. (see https://amp.dev/documentation/components/amp-script) [AMP_TAG_PROBLEM]
```
Comment (kristoferbaxter) [inline extensions/amp-script/0.1/test/validator-amp-script-not-origin-trial.out]: Intentional change?

File: extensions/amp-script/amp-script.md [hunk 3]
```diff
@@ -99,7 +106,7 @@ Overall mutation rules are:
 
 1. Mutations are always accepted after a user action for a user action interval of 5 seconds.
 2. The 5 seconds interval is extended if the user script performs a `fetch()` operation.
-3. Smaller `amp-script` elements with height under `300px` and non-`container` layout are allowed unlimitted mutations.
+3. Smaller `amp-script` elements with height under `300px` and non-`container` layout are allowed unlimited mutations.
```
Comment (kristoferbaxter) [inline extensions/amp-script/amp-script.md]: Nice catch.

File: extensions/amp-script/0.1/test/validator-amp-script-not-origin-trial.out (hunk 2 above)
Comment (choumx) [reply extensions/amp-script/0.1/test/validator-amp-script-not-origin-trial.out]: Hmm this was auto-generated by `gulp validator --update_tests`. I think it's harmless.

File: extensions/amp-script/validator-amp-script.protoascii (hunk 1 above)
[the remark this answers is not in the document]
Comment (choumx) [reply extensions/amp-script/validator-amp-script.protoascii]: Changed to use `[type=text/plain][target=amp-script]` to preempt the same question from users and be more consistent with amp-mustache.

**Comment by twifkak** [comment]
This affects how https://github.com/ampproject/amppackager#security-considerations are applied. I'm going to send a change to amppkg not to sign pages with inline amp-scripts, unless I hear a compelling alternative. I was hoping to have time to make a less rushed decision re: this.

**Pull request merged**